vnext/Desktop/Modules/BlobModule.cpp

```diff
@@ -0,0 +1,239 @@
+            auto blobId = blob["blobId"].asString();
+
+            auto bufferPart = m_contentHandler->ResolveMessage(blob["blobId"].asString(), blob["offset"].asInt(), blob["size"].asInt());
+            buffer.reserve(buffer.size() + bufferPart.size());
+            buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
+          }
+          else if (type == "string")
+          {
+            auto data = part["data"].asString();
+            auto bufferPart = vector<uint8_t>(data.begin(), data.end());
+
+            buffer.reserve(buffer.size() + bufferPart.size());
+            buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
+          }
+          else// if (auto instance = getInstance().lock())
+          {
+            //TODO: Error?
+            //instance->callJSFunction("RCTDeviceEventEmitter", "emit", dynamic::array("eventName", dynamic{}));
+            return;
+          }
+
+          m_contentHandler->StoreMessage(std::move(buffer), std::move(blobId));
+        }
+      }
+    ),
+
+    Method(
+      "release",
+      [this](dynamic args)  // blobId: string
+      {
+        auto blobId = jsArgAsString(args, 0);
+
+        m_contentHandler->RemoveMessage(std::move(blobId));
+      }
+    )
+  };
+}
+// clang-format on
+
+#pragma endregion CxxModule overrides
+
+#pragma region IWebSocketModuleContentHandler overrides
+
+void BlobWebSocketModuleContentHandler::Register(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  m_socketIDs.insert(socketID);
+}
+
+void BlobWebSocketModuleContentHandler::Unregister(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  if (m_socketIDs.find(socketID) != m_socketIDs.end())
+    m_socketIDs.erase(socketID);
+}
+
+bool BlobWebSocketModuleContentHandler::IsRegistered(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  return m_socketIDs.find(socketID) != m_socketIDs.end();
+}
+
+vector<uint8_t>
+BlobWebSocketModuleContentHandler::ResolveMessage(string &&blobId, int64_t offset, int64_t size) noexcept
+/*override*/
+{
+  lock_guard<mutex> lock{m_blobsMutex};
+
+  auto data = m_blobs.at(std::move(blobId));
+  auto start = data.cbegin() + static_cast<size_t>(offset);
+  auto end = start + static_cast<size_t>(size);
+
+  return vector(start, end);
+}
+
+void BlobWebSocketModuleContentHandler::RemoveMessage(string &&blobId) noexcept {
+  lock_guard<mutex> lock{m_blobsMutex};
+
+  m_blobs.erase(std::move(blobId));
+}
+
+void BlobWebSocketModuleContentHandler::ProcessMessage(string &&message, dynamic &params) /*override*/
+{
+  params["data"] = std::move(message);
+}
+
+void BlobWebSocketModuleContentHandler::ProcessMessage(vector<uint8_t> &&message, dynamic &params) /*override*/
+{
+  auto blob = dynamic::object();
+  blob("offset", 0);
+  blob("size", message.size());
+
+  // substr(1, 36) strips curly braces from a GUID.
```

Comment:
>substr [](start = 5, length = 6)

Just for my education: 
I'm missing how the blobId is made accessible for the other api calls...
I see it created here, stored in m_blobs, but nowhere is it extracted... only used as key to erase it from the set. 

I'm basically missing how the JS side is getting access to the id. I would have expected `blob("blobID", blobId)` here to add it to the dynamic ojbect as it is read from the JS side on line: 124....

Reply:
The JS side doesn't access the blob ID.

It's generated and passed internally; an implementation detail.

Reply:
I'm still confused :D
Could you show me where is the BlobId that is created here used?
As mentioned it is passed to `StoreMessage` which` then uses it on `m_blobs`. Looking at all usages of m_blobs, I didn't see any code that extracts or uses the blobId. 
I thought the code above that dynamically extract it from the object in `getMethods` were using it to look it back up...

-or- is there another place where blobId's are generated taht I missed?



---
In reply to: [494041779](https://github.com/microsoft/react-native-windows/pull/6079#discussion_r494041779) [](ancestors = 494041779)

Reply:
See `createFromParts`:

```
              "createFromParts",
              [this](dynamic args) {
                auto parts = jsArgAsArray(args, 0); // Array<Object>
                auto blobId = jsArgAsString(args, 1);
                vector<uint8_t> buffer{};

                for (const auto &part : parts) {
                  auto type = part["type"];
                  if (type == "blob") {
                    auto blob = part["data"];
                    auto bufferPart = m_contentHandler->ResolveMessage(
                        blob["blobId"].asString(), blob["offset"].asInt(), blob["size"].asInt());
                    buffer.reserve(buffer.size() + bufferPart.size());
                    buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
                  } else if (type == "string") {
   ...
```

When a blob is to stored as multiple fragments (parts), each fragment will itself be a blob with an ID.
The ID is used to retrieve the actual binary contents from the native object (in this implementation, that would be `BlobWebSocketModuleContentHandler`).

Hope this makes it clearer? 😅